src/partition/repart.c

```diff
@@ -3349,38 +3724,25 @@ static int partition_populate_directory(Partition *p, char **ret_root, char **re
          * tree beforehand where we merge all our inputs. We then use this merged source tree to create the
          * read-only filesystem. */
 
-        if (!fstype_is_ro(p->format)) {
+        if (!mkfs_supports_root_option(p->format) || (strv_isempty(p->copy_files) && strv_isempty(p->make_directories))) {
                 *ret_root = NULL;
                 *ret_tmp_root = NULL;
                 return 0;
         }
 
-        /* If we only have a single directory that's meant to become the root directory of the filesystem,
-         * we can shortcut this function and just use that directory as the root directory instead. If we
-         * allocate a temporary directory, it's stored in "ret_tmp_root" to indicate it should be removed.
-         * Otherwise, we return the directory to use in "root" to indicate it should not be removed. */
-
-        if (strv_length(p->copy_files) == 2 && strv_length(p->make_directories) == 0 && streq(p->copy_files[1], "/")) {
-                _cleanup_free_ char *s = NULL;
-
-                r = chase_symlinks(p->copy_files[0], arg_root, CHASE_PREFIX_ROOT, &s, NULL);
-                if (r < 0)
-                        return log_error_errno(r, "Failed to resolve source '%s%s': %m", strempty(arg_root), p->copy_files[0]);
-
-                *ret_root = TAKE_PTR(s);
-                *ret_tmp_root = NULL;
-                return 0;
-        }
-
         r = mkdtemp_malloc("/var/tmp/repart-XXXXXX", &root);
```

Comment:
Temporary directories are created with 700 permissions. Since `mksquashfs` copies the mode of the root directory the whole file system is only accessible to root. In case of the /usr file system this breaks a few things.

There are three ways this can be fixed, I think:

1. Chmod the temporary directory to a more permissive mode (from the `CopyFiles=` source?)
2. Create a subdirectory within the temporary directory and use that as the filesystem root.
3. Use [mksquashfs -root-mode option](https://man.archlinux.org/man/extra/squashfs-tools/mksquashfs.1.en#root~2)

Reply:
We now chown the temporary directory to 0755. I expect this to be a temporary fix though as we need a better mechanism to apply modes from CopyFiles= sources but we need to think this through a little bit as users don't always intend to copy directory modes when using CopyFiles=. #25051 has some more discussion surrounding this problem.